src/gradient/gradient.class.ts

```diff
@@ -1,493 +1,342 @@
+import { parseTransformAttribute } from "../parser/parseTransformAttribute";
+import { matrixToSVG, populateWithProperties } from "../util";
+import { parseColorStops, parseCoords, parseGradientUnits, parseType } from "./parser";
+import { ColorStop, GradientCoords, GradientOptions, GradientType, GradientUnits, SVGOptions } from "./typedefs";
+
+/**
+ * @todo remove this transient junk
+ */
+type FabricObject = any;
+
+/**
+ * Gradient class
+ * @class Gradient
+ * @tutorial {@link http://fabricjs.com/fabric-intro-part-2#gradients}
+ */
+export class Gradient<S, T extends GradientType = S extends GradientType ? S : 'linear'> {
 
-(function(global) {
-  var fabric = global.fabric;
-  /* _FROM_SVG_START_ */
-  function getColorStop(el, multiplier) {
-    var style = el.getAttribute('style'),
-        offset = el.getAttribute('offset') || 0,
-        color, colorAlpha, opacity, i;
+  /**
+   * Horizontal offset for aligning gradients coming from SVG when outside pathgroups
+   * @type Number
+   * @default 0
+   */
+  offsetX
 
-    // convert percents to absolute values
-    offset = parseFloat(offset) / (/%$/.test(offset) ? 100 : 1);
-    offset = offset < 0 ? 0 : offset > 1 ? 1 : offset;
-    if (style) {
-      var keyValuePairs = style.split(/\s*;\s*/);
+  /**
+   * Vertical offset for aligning gradients coming from SVG when outside pathgroups
+   * @type Number
+   * @default 0
+   */
+  offsetY
 
-      if (keyValuePairs[keyValuePairs.length - 1] === '') {
-        keyValuePairs.pop();
-      }
+  /**
+   * A transform matrix to apply to the gradient before painting.
+   * Imported from svg gradients, is not applied with the current transform in the center.
+   * Before this transform is applied, the origin point is at the top left corner of the object
+   * plus the addition of offsetY and offsetX.
+   * @type Number[]
+   * @default null
+   */
+  gradientTransform: number[] | null = null
 
-      for (i = keyValuePairs.length; i--; ) {
+  /**
+   * coordinates units for coords.
+   * If `pixels`, the number of coords are in the same unit of width / height.
+   * If set as `percentage` the coords are still a number, but 1 means 100% of width
+   * for the X and 100% of the height for the y. It can be bigger than 1 and negative.
+   * allowed values pixels or percentage.
+   * @type GradientUnits
+   * @default 'pixels'
+   */
+  gradientUnits: GradientUnits
 
-        var split = keyValuePairs[i].split(/\s*:\s*/),
-            key = split[0].trim(),
-            value = split[1].trim();
+  /**
+   * Gradient type linear or radial
+   * @type GradientType
+   * @default 'linear'
+   */
+  type: T
+
+  coords: GradientCoords<T>
+
+  colorStops: ColorStop[]
+
+  private id: string | number
+
+  constructor({
+    type = 'linear' as T,
+    gradientUnits = 'pixels',
+    coords,
+    colorStops = [],
+    offsetX = 0,
+    offsetY = 0,
+    gradientTransform,
+    id
+  }: GradientOptions<T>) {
+    const uid = fabric.Object.__uid++;
```

Comment:
in general, this part has been removed
```
      // sets everything, then coords and colorstops get sets again
      Object.keys(options).forEach(function(option) {
        _this[option] = options[option];
      });
```

This is something we do often with shapes.
Is TS or classes going in our way here?

Reply:
It can be done
In shapes it is done in `set` only right?

Reply:
i think is in the basic setOptions of object.
Is not a big deal for gradients because gradients are often just used for fills and not to carry any logic, but sometimes devs add random properties to object without subclassing